freedomofpress/securedrop-workstation
# Skips running the updater if last update was good, a reboot isn't needed, and a config-specified delta isn't defined.

## Status

Ready for review 

## Description of Changes

Fixes #402.
- updates the SecureDrop desktop shortcut executable to include a parameter `--skip-delta=<update_skip_delta>` 
- updates the Securedrop launcher  to add the optional `--skip-delta` parameter above. If the last update was successful and less than `update_skip_delta` seconds ago, then the launcher runs the SecureDrop client directly. Otherwise, it runs the updater. if `--skip-delta` is omitted, a default interval of 8 hours is assumed. If `--skip-delta` is 0, the updater always runs.

## Testing
On this branch:
- install the SecureDrop Workstation (remembering to set the `update_skip_delta` field in config.json)
- confirm that the SecureDrop desktop shortcut's command is now `/opt/securedrop/launcher/sdw-launcher.py --skip-delta=28800`
- delete the dom0 directory `~/.securedrop_launcher` if it exists

### testing initial and second run
- double-click the SecureDrop shortcut
  - [x] confirm that the updater is launched
  - [x] confirm that the logfile at `~/.securedrop_launcher/logs/launcher.log` contains the line:
  ```
  INFO: Update status not available, launching updater.
  ```
- complete the update check, applying updates and rebooting if necessary. If a reboot is not necessary, click **Continue** to launch the client, then close the client
- double-click the SecureDrop shortcut
  - [x] confirm that the client is launched directly without the updater starting
  - [x] confirm that the logfile above contains the line:
  ```
  INFO: Updates OK and interval not expired, launching client.
  ```
- close the client

### testing updates_required status
- update the dom0 JSON file `~/.securedrop_launcher/sdw-update-status`, changing the value of the `status` field to `"1"`.
- double-click the SecureDrop shortcut
  - [x] confirm that the updater is launched
  - [x] confirm that the logfile at `~/.securedrop_launcher/logs/launcher.log` contains the line:
  ```
  INFO: Update status is 1, launching updater.
  ```
- after updates are complete, click **Continue**, then close the client

### testing reboot_required status (without reboot having been performed)
- update the dom0 JSON file `~/.securedrop_launcher/sdw-update-status`, changing the value of the `status` field to `"2"`.
- double-click the SecureDrop shortcut
  - [x] confirm that the updater is launched
  - [x] confirm that the logfile at `~/.securedrop_launcher/logs/launcher.log` contains the line:
  ```
  INFO: Required reboot pending, launching updater
  ```
- after updates are complete, click **Continue**, then close the client

### testing updates_failed status
- update the dom0 JSON file `~/.securedrop_launcher/sdw-update-status`, changing the value of the `status` field to `"3"`.
- double-click the SecureDrop shortcut
  - [x] confirm that the updater is launched
  - [x] confirm that the logfile at `~/.securedrop_launcher/logs/launcher.log` contains the line:
  ```
  INFO: Update status is 3, launching updater.
  ```
- after updates are complete, click **Reboot**


### testing reboot_required status (with required reboot)
- update the dom0 JSON file `~/.securedrop_launcher/sdw-update-status`, changing the value of the `status` field to `"2"`.
- reboot Qubes.
- double-click the SecureDrop shortcut
  - [x] confirm that the client is launched
  - [x] confirm that the logfile at `~/.securedrop_launcher/logs/launcher.log` contains the line:
  ```
  INFO: Required reboot performed, updating status and launching client..
  ```
- [x] confirm that the `sdw-update-status` file has been updated with status `0` and a new timestamp
- close the client

### testing interval expiry
- open a dom0 Terminal
- run the command `/opt/securedrop/launcher/sdw-launcher.py`
  - [x] confirm that the client starts without the updater running
- run the command `/opt/securedrop/launcher/sdw-launcher.py --skip-delta=0`
  - [x] confirm that the updater runs directly
- once the update completes, click **Continue**, then close the client
- in the Terminal, run the command `/opt/securedrop/launcher/sdw-launcher.py --skip-delta=600`
  - [x] confirm that the client starts without the updater running
- wait a little over 10 seconds, then, in the Terminal, run the command `/opt/securedrop/launcher/sdw-launcher.py --skip-delta=10`
  - [x] confirm that the updater runs.

## Checklist

### If you have made changes to the provisioning logic

- [x] Linting (`make flake8`) and tests (`make test`) pass in dom0 of a Qubes install
**`make test` fails an unrelated qubes-rpc test due to system setup**

- [x] I have added/removed files, and have updated packaging logic in MANIFEST.in and rpm-build/SPECS/securedrop-workstation-dom0-config.spec
